### ISSUE TITLE
Implementation difference in normalization(shark) and zscore(MATLAB)

### ISSUE THREAD

**Issue by DetailFreak** [Shark-ML/Shark#191]
I was testing normalization using:

A: The default shark normalization:
`bool removeMean = true;
	Normalizer<RealVector> normalizer;
	NormalizeComponentsUnitVariance<RealVector> normalizingTrainer(removeMean);
	normalizingTrainer.train(normalizer, data.inputs());
	LabeledData<RealVector, unsigned int> normalizedData = transformInputs(data, normalizer);`

B: MATLAB's function:
`A = zscore(A)`

Both make mean 0 and stddev 1, yet produce difference in output.

I was follwing this this tutorial:[Support Vector Machines: Model Selection Using Cross-Validation and Grid-Search](http://image.diku.dk/shark/sphinx_pages/build/html/rest_sources/tutorials/algorithms/svmModelSelection.html)

Using 10 fold crossvalidation on a dataset of 304x24 features I am getting a best accuracy of 83% for A and 96.6% for B.
Please let me know of the difference in their implementation.

**Comment by Ulfgard** [comment]
I am unable to follow you. Have you verified that the datasets produces after normalization are indeed diferent? What did you do afterwards? have you used the matlab normalized data and imported that in shark or did you use the matlab statistical toolbox? If yes, have you ensured that the parameters found are the same when not using normalization? 

Are the numbers you show a test error? If yes, have you ensured with matlab that you use the same parameters to normalize the test set as you used to normalize the training set? (you can not just run zscore on the test set and this alone could explain the differences in performance). Have you compared training errors as well?

Finally: is there something weird about your dataset, like missing data or numerically unstable value ranges?

The easiest way to proof that this is indeed the difference is that you simply wite out both datasets after normalization (or just the first few points) and compare them.

**Comment by DetailFreak** [comment]
Yes, I did check the difference in the datasets after normalization. I imported the resultant dataset from MATLAB to shark. 

I normalized my whole dataset at once as I'm using crossvalidation. The training errors decreased too.  

My dataset is fine. 
I did find the difference between matlab and shark's normalized data in matlab to confirm my doubt.

**Comment by Ulfgard** [comment]
Cool. Thank you for this information. I do not have access to your data, so I am going to need a bit more of information to help me see where something might be off (as our unit tests run fine...) Could you run this after the normalization? 

```
RealVector mean;
RealVector variance;
meanvar(data.inputs(), mean, variance);
std::cout<<mean<<variance;
meanvar(normalizedData, mean, variance);
std::cout<<mean<<variance;
```

**Comment by DetailFreak** [comment]
`[24](0.539699,3.94704,0.18549,0.108891,6.13557,1.39583,0.833436,5.55739,0.185825,0.106977,4,0,1.04344,5.40928,0.178173,0.105958,2.65355,0.518668,112.177,41.2422,-3.19338e+302,1.75252e+36,-1.6589e+303,2.92455e+78)[24](0.0604657,6.01701,6.44899e-05,3.75232e-05,0.0777707,0.0804438,0.151289,8.70893,2.15645e-05,9.17141e-06,0,0,0.213404,5.11802,2.75454e-05,7.81268e-06,0.00643106,0.0021071,6.52935,6.50955,inf,9.30609e+74,inf,2.59156e+159)[24](2.62948e-17,-6.5883e-16,1.98671e-15,4.72137e-15,2.13864e-15,-4.73306e-16,3.62283e-16,-4.55776e-16,-7.75987e-15,-1.03075e-14,0,0,-1.63612e-16,-1.40239e-16,-5.45762e-15,8.27408e-15,2.47755e-15,4.07277e-15,1.10321e-14,-4.96679e-15,0,-8.18059e-17,0,1.22709e-16)[24](1,1,1,1,1,1,1,1,1,1,0,0,1,1,1,1,1,1,1,1,0,1,0,1)`

**Comment by Ulfgard** [comment]
Hi, 

your dataset is not fine. You have numerically *extremely* unstable value ranges, see the last values: of the means
-3.19338e+302,1.75252e+36,-1.6589e+303,2.92455e+78

Two of those are extremely close to DBL_MAX, and are 300 orders of magnitude larger than other values you report.

I suspect the reason why matlab works is that it first computes the mean and then subtracts it from all entries. this will most likely push some of the columns to be completely 0, which leads to the variance being 0 which is guarded against. Shark computes the square of the entrie firsts, which means that e+303 is overflowing and becoming DBL_INF, which makes the variance inf. Both, inf and 0 are obviously wrong, where inf is closer to what i would consider the "truth" in this case.

Please, fix your dataset. I suspect this is caused by missing data/nan/invalid-values which are at some point replaced by DBL_MAX or similar.

I am closing this as invalid. If you have still questions, don't hesitate, though.
